internal/indexer/layerscanner/layerscanner.go

```diff
@@ -83,100 +55,128 @@ func (ls *layerScanner) Scan(ctx context.Context, manifest claircore.Digest, lay
+	filter := func(s indexer.VersionedScanner) bool {
+		n := s.Name()
+		var cfgMap map[string]func(interface{}) error
+		switch k := s.Kind(); k {
+		case "package":
+			cfgMap = ls.Opts.ScannerConfig.Package
+		case "repository":
+			cfgMap = ls.Opts.ScannerConfig.Repo
+		case "distribution":
+			cfgMap = ls.Opts.ScannerConfig.Dist
+		default:
+			log.Warn().
+				Str("kind", k).
+				Str("scanner", n).
+				Msg("unknown scanner kind")
+			return true
 		}
 
-		for _, s := range rs {
-			ss := s
-			g.Go(func() error {
-				return ls.scanRepos(gctx, ll, ss)
-			})
+		if f, ok := cfgMap[n]; ok {
+			cs, csOK := s.(indexer.ConfigurableScanner)
+			rs, rsOK := s.(indexer.RPCScanner)
+			switch {
+			case !csOK && !rsOK:
+				log.Warn().
+					Str("scanner", n).
+					Msg("configuration present for an unconfigurable scanner, skipping")
+			case csOK && rsOK:
+				fallthrough
+			case !csOK && rsOK:
+				if err := rs.Configure(ctx, f, ls.Opts.Client); err != nil {
+					log.Error().
+						Str("scanner", n).
+						Err(err).
+						Msg("configuration failed")
+					return true
+				}
+			case csOK && !rsOK:
+				if err := cs.Configure(ctx, f); err != nil {
+					log.Error().
+						Str("scanner", n).
+						Err(err).
+						Msg("configuration failed")
+					return true
+				}
+			}
 		}
+		return false
 	}
 
-	if err := g.Wait(); err != nil {
-		return err
-	}
-
-	return nil
-}
-
-func (ls *layerScanner) scanPackages(ctx context.Context, layer *claircore.Layer, s indexer.PackageScanner) error {
-	log := zerolog.Ctx(ctx).With().
-		Str("component", "internal/indexer/layerscannner/layerScanner.scanPackages").
-		Str("scanner", s.Name()).
-		Str("layer", layer.Hash.String()).
-		Logger()
-
-	log.Debug().Msg("starting package scan")
-	if err := ls.addToken(ctx); err != nil {
-		return err
-	}
-	defer ls.discardToken()
-
-	ok, err := ls.Store.LayerScanned(ctx, layer.Hash, s)
-	if err != nil {
-		return err
-	}
-	if ok {
-		log.Debug().Msg("layer already scanned")
-		return nil
+	// These loops do the configure & filter. This is done now to avoid doing it
+	// for every layer.
+	var i int
+	i = 0
+	for _, s := range ps {
+		if !filter(s) {
+			ps[i] = s
+			i++
+		}
 	}
-
-	v, err := s.Scan(ctx, layer)
-	if err != nil {
-		return fmt.Errorf("scanner: %v error: %v", s.Name(), err)
+	ps = ps[:i]
+	i = 0
+	for _, s := range rs {
+		if !filter(s) {
+			rs[i] = s
+			i++
+		}
 	}
-	err = ls.Store.SetLayerScanned(ctx, layer.Hash, s)
-	if err != nil {
-		return fmt.Errorf("could not set layer scanned: %v", layer)
+	rs = rs[:i]
+	i = 0
+	for _, s := range ds {
+		if !filter(s) {
+			ds[i] = s
+			i++
+		}
 	}
-
-	if v == nil {
-		log.Debug().Msg("scan returned a nil")
-		return nil
+	ds = ds[:i]
+
+	sem := semaphore.NewWeighted(int64(ls.cLevel))
+	g, ctx := errgroup.WithContext(ctx)
+	// Scan is a closure to capture the values in the loops and then call the
+	// scan method.
+	scan := func(l *claircore.Layer, s indexer.VersionedScanner) func() error {
```

Comment:
can you avoid this scan variable shadow with ls.scan?
put the sem on the layerScanner struct and make it sharable and then use it in ls.scan
seems a bit silly to have ls.scan and then a scan closure just to capture a sem that can be shared in he ls.scan method anyway.

Reply:
It's not to capture the semaphore, it's to capture the loop variables.

Reply:
can you make it so the process of performing a scan is self contained in layerScanner.Scan?
its simpler then trying to hop around between the Scan function, the scan variable, the function that is returned by the scan variable, and the ls.scan function.